cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -543,20 +543,32 @@ public void saveProgramRuntimeArgs(HttpRequest request, HttpResponder responder,
   }
 
   @GET
-  @Path("/apps/{app-id}/{program-type}/{program-id}")
+  @Path("/apps/{app-name}/{program-type}/{program-name}")
   public void programSpecification(HttpRequest request, HttpResponder responder,
-                                   @PathParam("namespace-id") String namespaceId, @PathParam("app-id") String appId,
+                                   @PathParam("namespace-id") String namespaceId, @PathParam("app-name") String appName,
                                    @PathParam("program-type") String type,
-                                   @PathParam("program-id") String programId) throws Exception {
+                                   @PathParam("program-name") String programName) throws Exception {
+    programSpecification(request, responder, namespaceId, appName, ApplicationId.DEFAULT_VERSION, type, programName);
+  }
+
+
+  @GET
+  @Path("/apps/{app-name}/versions/{app-version}/{program-type}/{program-name}")
+  public void programSpecification(HttpRequest request, HttpResponder responder,
+                                   @PathParam("namespace-id") String namespaceId, @PathParam("app-name") String appName,
+                                   @PathParam("app-version") String appVersion,
+                                   @PathParam("program-type") String type,
+                                   @PathParam("program-name") String programName) throws Exception {
     ProgramType programType = getProgramType(type);
     if (programType == null) {
       throw new MethodNotAllowedException(request.getMethod(), request.getUri());
     }
 
-    ProgramId id = new ProgramId(namespaceId, appId, programType, programId);;
+    ApplicationId application = new ApplicationId(namespaceId, appName, appVersion);
+    ProgramId id = application.program(programType, programName);
     ProgramSpecification specification = lifecycleService.getProgramSpecification(id);
     if (specification == null) {
-      throw new NotFoundException(programId);
+      throw new NotFoundException(programName);
```

Comment:
Ok. It just seems strange to pass in a `String` value, instead of the `ProgramId` object.
The exception message differs in either of those cases.